File: web-client/src/app/blog/blog.service.ts
import {Observable} from "rxjs/Observable"
import {BlogEntry, BlogEntryHeader, copyBlogEntry, NewBlogEntry} from "./blog.models"
import "rxjs/add/observable/of"

export interface BlogService {

  getBlogEntryHeaders(): Observable<BlogEntryHeader[]>

  getBlogEntryById(id: number): Observable<BlogEntry>

  createBlogEntry(newBlogEntry: NewBlogEntry): Observable<BlogEntry>

}


export class StubBlogService implements BlogService {

  blogEntries: BlogEntry[] = []
  private idCounter = 0

  constructor() {
    for (let i = 0; i < 3; i++) {
      this.idCounter++
      this.blogEntries.push(
        new BlogEntry(
          this.idCounter,
          new Date(),
          new Date(),
          new Date(),
          "My Blog Entry " + this.idCounter,
          "A bunch of text that gets shown as a blog entry. " +
          "A bunch of text that gets shown as a blog entry. " +
          "A bunch of text that gets shown as a blog entry. " +
          "A bunch of text that gets shown as a blog entry. " +
          "A bunch of text that gets shown as a blog entry. " +
          "A bunch of text that gets shown as a blog entry. " +
          "A bunch of text that gets shown as a blog entry. " +
          "A bunch of text that gets shown as a blog entry. "
        )
      )
    }
  }

  createBlogEntry(newBlogEntry: NewBlogEntry): Observable<BlogEntry> {
    const created = new BlogEntry(
      this.nextID(),
      newBlogEntry.publish ? new Date() : null,
      new Date(),
      new Date(),
      newBlogEntry.title,
      newBlogEntry.content
    )
    this.blogEntries.push(created)
    return Observable.of(copyBlogEntry(created))
  }

  getBlogEntryHeaders(): Observable<BlogEntryHeader[]> {
    return Observable.of(this.blogEntries.map((entry) => {
      return entry.toHeader()
    }))
  }

  getBlogEntryById(id: number): Observable<BlogEntry> {
    for (const entry of this.blogEntries) {
      if (entry.id === id) {
        return Observable.of(entry)
      }
    }
    return Observable.of(null)
  }

  private nextID(): number {
    this.idCounter++
    return this.idCounter
  }

}

File: web-server/src/main/kotlin/com/binarymonks/gonzo/core/blog/service/BlogService.kt
package com.binarymonks.gonzo.core.blog.service

import com.binarymonks.gonzo.core.blog.api.*
import com.binarymonks.gonzo.core.blog.persistence.BlogEntryDraftEntity
import com.binarymonks.gonzo.core.blog.persistence.BlogEntryPublished
import com.binarymonks.gonzo.core.blog.persistence.BlogRepo
import com.binarymonks.gonzo.core.time.nowUTC
import com.binarymonks.gonzo.core.users.persistence.UserRepo
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service

@Service
class BlogService : Blog {

    @Autowired
    lateinit var blogRepo: BlogRepo

    @Autowired
    lateinit var userRepo: UserRepo


    override fun createBlogEntry(blogEntryNew: BlogEntryNew): BlogEntryDraft {

        return blogRepo.save(BlogEntryDraftEntity(
                title = blogEntryNew.title,
                content = blogEntryNew.content,
                author = userRepo.findById(blogEntryNew.authorID).get(),
                created = nowUTC(),
                updated = nowUTC()
        )).toBlogEntryDraft()
    }


    override fun updateBlogEntry(update: BlogEntryUpdate): BlogEntryDraft {
        val entry = blogRepo.findById(update.id).get()
        val changed = listOf(
                entry.publishedBlog?.content != update.content,
                entry.publishedBlog?.title != update.title
        ).any { it }
        if (changed) {
            entry.title = update.title
            entry.content = update.content
            entry.unpublishedChanges = true
            entry.updated = nowUTC()
        }
        return blogRepo.save(entry).toBlogEntryDraft()
    }

    override fun getBlogEntryDraftByID(blogID: Long) = blogRepo.findById(blogID).get().toBlogEntryDraft()

    override fun getBlogEntryDraftHeaders(authorID: Long): List<BlogEntryHeader> {
        val userEntity = userRepo.findById(authorID).get()
        return blogRepo.findAllByAuthor(userEntity).map { it.toBlogEntryDraft().toHeader() }
    }

    override fun getBlogEntryHeadersByAuthor(authorID: Long): List<BlogEntryHeader> {
        val userEntity = userRepo.findById(authorID).get()
        return blogRepo.findAllByAuthor(userEntity).filter {
            it.publishedBlog!=null }.map {
            it.toBlogEntryDraft().toHeader() }
    }

    override fun publishBlogEntry(blogID: Long) {
        val now = nowUTC()
        val entity = blogRepo.findById(blogID).get()
        if (entity.publishedBlog != null) {
            entity.publishedBlog!!.content = entity.content
            entity.publishedBlog!!.title = entity.title
            entity.publishedBlog!!.updated = now
        } else {
            entity.publishedBlog = BlogEntryPublished(
                    blogEntryDraftID = entity.id!!,
                    title = entity.title,
                    content = entity.content,
                    created = now,
                    updated = now
            )
        }
        entity.unpublishedChanges = false
        blogRepo.save(entity)
    }

    override fun getBlogEntryHeaders(): List<BlogEntryHeader> = blogRepo.findAll().filter {
        it.publishedBlog!=null }.map {
        it.toBlogEntryDraft().toHeader() }

    override fun getBlogEntryById(id: Long): BlogEntry = blogRepo.findById(id).get().toBlogEntry()

}
